### PR TITLE
Add a callback to the Push method of the deserializer to allow the client to reclaim memory early

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: base/push_deserializer_test.cpp [hunk 1]
```diff
@@ -170,5 +183,28 @@ TEST_F(PushDeserializerTest, SerializationDeserialization) {
...
+  auto serialized_trajectory =
+      std::make_unique<std::uint8_t[]>(trajectory_.ByteSize());
+  trajectory_.SerializePartialToArray(&serialized_trajectory[0],
+                                      trajectory_.ByteSize());
+  push_deserializer_->Start(&read_trajectory);
+  int left = trajectory_.ByteSize();
+  for (int i = 0; i < trajectory_.ByteSize(); i += kStompChunk) {
+    Bytes bytes(&serialized_trajectory[i], std::min(left, kStompChunk));
+    push_deserializer_->Push(bytes,
+                              std::bind(&PushDeserializerTest::Stomp,
```
Comment (eggrobin) [inline base/push_deserializer_test.cpp]: bad indent.

File: base/push_deserializer_test.cpp [hunk 2]
```diff
@@ -59,6 +60,12 @@ class PushDeserializerTest : public ::testing::Test {
     }
   }
 
+  static void Stomp(const Bytes& bytes) {
```
Comment (eggrobin) [inline base/push_deserializer_test.cpp]: const&

File: base/push_deserializer_test.cpp [hunk 3]
```diff
@@ -59,6 +60,12 @@ class PushDeserializerTest : public ::testing::Test {
...
+    for (int i = 0; i < bytes.size; ++i) {
+      bytes.data[i] = 0xCD;
+    }
```
Comment (eggrobin) [inline base/push_deserializer_test.cpp]: why not `memset(bytes.data, 0xCD, bytes.size);`?

Comment (pleroy) [inline base/push_deserializer_test.cpp]: Done.

File: base/push_deserializer_test.cpp (hunk 2 above)
Comment (pleroy) [inline base/push_deserializer_test.cpp]: Done.

File: base/push_deserializer_test.cpp (hunk 1 above)
Comment (pleroy) [inline base/push_deserializer_test.cpp]: Done.

**Pull request merged**